src/dataframe/dataframe.jl

```diff
@@ -1827,27 +1873,25 @@ function Base.push!(df::DataFrame, row::Any; promote::Bool=false)
     try
         for (i, (col, val)) in enumerate(zip(_columns(df), row))
             current_col += 1
+            @assert length(col) == nrows
             S = typeof(val)
             T = eltype(col)
             if S <: T || !promote || promote_type(S, T) <: T
-                push!(col, val)
+                insert!(col, loc, val)
             else
                 newcol = Tables.allocatecolumn(promote_type(S, T), targetrows)
-                copyto!(newcol, 1, col, 1, nrows)
-                newcol[end] = val
                 firstindex(newcol) != 1 && _onebased_check_error()
+                copyto!(newcol, 1, col, 1, loc-1)
+                copyto!(newcol, loc+1, col, loc, nrows-loc+1)
+                newcol[loc] = val
                 _columns(df)[i] = newcol
             end
         end
-        current_col = 0
-        for col in _columns(df)
-            current_col += 1
-            @assert length(col) == targetrows
-        end
     catch err
         #clean up partial row
         for col in _columns(df)
-            resize!(col, nrows)
+            length(col) == targetrows && deleteat!(col, loc)
+            @assert length(col) == nrows
```

Comment:
I imagine the new code could give weird results in a corner case where several columns at the end of the data frame already had the target size (but not others). Or if an unrelated error happens when adding a row to such a data frame. Then `@assert` above will fail for the first one, and we will delete one entry at position `loc` in each of the other too long columns. So at the end the data frame won't be the same as before calling `insert!`. Not sure whether we should care or not.

Reply:
> I imagine the new code could give weird results in a corner case where several columns at the end of the data frame already had the target size (but not others).

In normal circumstances it is not possible. The reason is that in the loop above I fail fast with `@assert length(col) == nrows` check in the `try` block. Which means that we exit the loop the first time something wrong happens in processing of the data. In other words if we get an error in column e.g. 3 then columns 4 and later are not touched

> Or if an unrelated error happens when adding a row to such a data frame.

Indeed if some system-level error happens then things might be garbage. But in such a situation there is no way to recover from the error (as things must have failed on a very low level, so we anyway do not know how to recover). The code assumes that if things failed and we see a column into which a row was added then this addition happened at `loc` position.

> Then `@assert` above will fail for the first one, and we will delete one entry at position loc in each of the other too long columns.

This is not possible under "normal" work of Julia (i.e. if we do not hit some system level error). As commented above. If we have an error in column 3, later columns are not touched. The only case when they would be touched is when are aliases of e.g. column 3, but in this case we recover from error in column `3` and some later column is exactly the same as column 3 (as it is its alias).

Note that we cannot have a potential problem of overlapping memory segments that can happen in views because views are not resizable, so function will error early anyway.

---

Side note: if you compare this implementation to the `push!` implementation indeed it is more complex (and it has to be). The issue is that in `push!` we knew we have to trim all the vectors and did not have to "fail fast" on aliases. Now we need to fail fast on aliases , as otherwise indeed we would push multiple values at location `loc` and then we would not know how many to remove. In the current code we fail the first time we see something wrong in a column (so at most one incorrect insertion is made).

In short: I have thought about it when implementing the code. Things should be OK, but are not easy to reason about. We try to recover the proper state of the data frame as well as we can, but in general users should not rely on a fact that in-place operation properly recovers the state of the mutated object on error (Base Julia does not ensure this; we try, and I would say that most of the time we do it correctly).

Reply:
The situation I was thinking about is e.g. this one:
```julia
julia> df = DataFrame(x=1:3, y=1:3, z=1:3)
pus3×3 DataFrame
 Row │ x      y      z     
     │ Int64  Int64  Int64 
─────┼─────────────────────
   1 │     1      1      1
   2 │     2      2      2
   3 │     3      3      3

julia> push!(df.z, 4)
4-element Vector{Int64}:
 1
 2
 3
 4

julia> push!(df, (x=1, y="a", z=2))
┌ Error: Error adding value to column :y.
└ @ DataFrames ~/.julia/dev/DataFrames/src/dataframe/dataframe.jl:1713
ERROR: MethodError: Cannot `convert` an object of type String to an object of type Int64

julia> df
3×3 DataFrame
 Row │ x      y      z     
     │ Int64  Int64  Int64 
─────┼─────────────────────
   1 │     1      1      1
   2 │     2      2      2
   3 │     3      3      3
```
Note that the last value in column z was silently dropped, without signalling that the data frame was corrupt from the beginning. This is of course a rare situation, but the previous approach of keeping track of the index of the last column that was modified before the error happened was safer.

Reply:
> This is of course a rare situation, but the previous approach of keeping track of the index of the last column that was modified before the error happened was safer.

OK - I will make the "rollback" only affect the changed columns.